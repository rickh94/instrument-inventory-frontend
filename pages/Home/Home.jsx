import React from 'react'
import PropTypes from 'prop-types'
import {
  Typography,
  List,
  ListItem,
  withStyles,
  Paper,
  Link,
  ListItemIcon,
  ListItemText
} from '@material-ui/core'
import InputIcon from '@material-ui/icons/Input'
import ListIcon from '@material-ui/icons/FormatListBulleted'
import LabelIcon from '@material-ui/icons/Label'
import AddIcon from '@material-ui/icons/Add'
import { withRouter } from 'react-router-dom'

import { root } from '../../globalStyles'

const styles = {
  root,
  footer: {
    position: 'absolute',
    bottom: 0,
    display: 'flex',
    textAlign: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'center',
    width: '100%',
  },
  footerItem: {
    marginLeft: '0.2rem'
  }
}

function Home(props) {
  const { classes } = props
  return (
    <React.Fragment>
      <Paper className={classes.root}>
        <Typography variant="h4" color="inherit">
          Options
        </Typography>
        <List>
          <ListOptionItem
            to="/create"
            icon={<AddIcon />}
            text="Create a New Instrument"
          />
          <ListOptionItem
            to="/sign-out"
            icon={<LabelIcon />}
            text="Sign out instrument"
          />
          <ListOptionItem
            to="/retrieve-single"
            icon={<InputIcon />}
            text="Retrieve An Instrument"
          />
          <ListOptionItem
            to="/retrieve-multiple"
            icon={<ListIcon />}
            text="Retrieve Multiple Instruments"
          />
        </List>
      </Paper>
      <div className={classes.footer}>
        <Typography variant="body2" color="inherit" className={classes.footerItem}>
          &copy; Rick Henry 2019 |{' '}
        </Typography>
        <Link
          target="_blank"
          href="http://chittagongit.com/icon/violin-icon-27.html"
          rel="noopener"
        >
          <Typography variant="body2" color="inherit" className={classes.footerItem}>
            {' '}
            Icon Credit
          </Typography>
        </Link>
      </div>
    </React.Fragment>
  )
}

Home.propTypes = {
  classes: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired
}

export default withStyles(styles)(Home)

const ListOptionItem = withRouter(({ to, icon, text, history }) => {
  return (
    <ListItem button onClick={() => history.push(to)}>
      <ListItemIcon>{icon}</ListItemIcon>
      <ListItemText>{text}</ListItemText>
    </ListItem>
  )
})

ListOptionItem.propTypes = {
  to: PropTypes.string.isRequired,
  icon: PropTypes.element.isRequired,
  text: PropTypes.string.isRequired
}
