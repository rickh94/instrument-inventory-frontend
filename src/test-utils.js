/* eslint-disable react/prop-types */
import React from 'react'
import { MuiThemeProvider, createMuiTheme, indigo } from '@material-ui/core'
import { HelpersContext, SchemaContext } from './contexts'
import { MemoryRouter } from 'react-router-dom'
import { render } from '@testing-library/react'

const theme = createMuiTheme({
  palette: {
    primary: indigo,
    secondary: {
      main: '#0A81D1',
    },
  },
})

export const showAlert = jest.fn()
export const setSearchResults = jest.fn()
export const setFilterResults = jest.fn()
export const searchResults = jest.fn()
export const filterResults = jest.fn()

export const TestTheme = ({ children }) => (
  <MuiThemeProvider theme={theme}>{children}</MuiThemeProvider>
)

export const TestHelpers = ({ children }) => (
  <HelpersContext.Provider
    value={{
      showAlert,
      setSearchResults,
      setFilterResults,
      searchResults,
      filterResults,
    }}
  >
    {children}
  </HelpersContext.Provider>
)

export const TestRouter = ({ children }) => <MemoryRouter>{children}</MemoryRouter>

export const TestSchema = ({ children, schema = { components: { schemas: {} } } }) => (
  <SchemaContext.Provider value={schema}>{children}</SchemaContext.Provider>
)

export const TestEverything = ({ children }) => (
  <TestRouter>
    <TestTheme>
      <TestHelpers>{children}</TestHelpers>
    </TestTheme>
  </TestRouter>
)

// eslint-disable-next-line no-undef
export const flushPromises = () => new Promise(setImmediate)

const customRender = (ui, options, schema = { components: { schemas: {} } }) =>
  render(<TestSchema schema={schema}>{ui}</TestSchema>, {
    wrapper: TestEverything,
    ...options,
  })

export * from '@testing-library/react'

export { customRender as render }
