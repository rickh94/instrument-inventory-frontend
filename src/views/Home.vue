<template>
  <div class="max-w-4xl">
    <div class="m-5 shadow p-3 h-auto">
      <label for="number" class="text-gray-800 font-bold text-xl mb-1">Instrument Number or Student Name</label>
      <div class="flex items-start justify-between mt-2 mb-1">
        <div class="border-b focus-within:border-purple-300 border-gray-800 w-24 flex-grow flex items-center space-between">
          <input type="text"
                 name="number"
                 id="number"
                 class="appearance-none bg-transparent border-none w-full text-gray-800 mr-1 py-1 leading-tight focus:outline-none"
                 v-model="searchTerm"
                 @keyup.enter="onSubmit"
          >
          <button class="appearance-none" title="Scan Barcode" @click="scanner = true">
            <font-awesome-icon icon="barcode" class="mx-2"></font-awesome-icon>
          </button>
        </div>
        <button @click="onSubmit"
                class="appearance-none rounded ml-4 text-white font-bold"
                :class="searchTerm.length === 0 ? 'bg-gray-600' : 'bg-purple-600 hover:shadow hover:bg-purple-800'"
                :disabled="searchTerm.length === 0"
        >

          <pulse-loader v-if="loading" color="#fff" size="8" class="px-2 py-1"></pulse-loader>
          <span v-else class="inline-flex items-center px-2 py-1">
            <svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5 mr-1" viewBox="0 0 20 20" fill="currentColor">
              <path fill-rule="evenodd"
                    d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z"
                    clip-rule="evenodd" />
            </svg>
          Search
          </span>
        </button>
      </div>
      <div v-if="isAdmin">
        <label class="inline-flex items-center">
          <input type="checkbox" class="form-checkbox" v-model="showArchived">
          <span class="ml-2">Include Archived Instruments</span>
        </label>
      </div>
      <v-scanner @detected="detected" v-if="scanner" @close="scanner = false"></v-scanner>
    </div>
    <multiple-results v-show="searchResults.length > 1" />
    <v-modal v-if="showNotFound" @close="showNotFound = false" width-class="sm:max-w-lg">
      <h6 class="text-xl font-bold text-gray-900">Not Found</h6>
      <p class="text-gray-900">Could not find instrument {{ searchTerm }}. Would you like to create a new
        instrument?</p>
      <div class="flex justify-end mt-2 flex-row">
<!--        <button class="mx-1 appearance-none bg-red-600 text-white px-4 py-1 shadow rounded hover:bg-red-800 hover:shadow-lg"-->
<!--                @click="showNotFound = false">-->
<!--          Close-->
<!--        </button>-->
        <v-close-form-button @close="showNotFound = false" />
        <button class="flex items-center font-bold mx-1 appearance-none bg-green-600 text-white px-3 py-1 shadow rounded hover:bg-green-800 hover:shadow-lg"
                @click="beginCreateInstrument">
          <svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5 mr-1" viewBox="0 0 20 20" fill="currentColor">
            <path d="M17.414 2.586a2 2 0 00-2.828 0L7 10.172V13h2.828l7.586-7.586a2 2 0 000-2.828z" />
            <path fill-rule="evenodd" d="M2 6a2 2 0 012-2h4a1 1 0 010 2H4v10h10v-4a1 1 0 112 0v4a2 2 0 01-2 2H4a2 2 0 01-2-2V6z" clip-rule="evenodd" />
          </svg>
          Create
        </button>
      </div>
    </v-modal>
  </div>
</template>

<script>
import VScanner from "@/components/UI/VScanner";
import { API } from "aws-amplify";
import { mapMutations, mapState } from "vuex";
import checkAdmin from "@/mixins/checkAdmin";
import { PulseLoader } from "@saeris/vue-spinners";
import VCloseFormButton from "@/components/UI/buttons/VCloseFormButton";

function getPath(input) {
  // This regex will match the instrument number format and search for an instrument number.
  // If it does not match, it is assumed to be a name and searches in assigned and history.
  return input.match(/\w?\w?\d*-\d+/) ? "search/number" : "search/assigned-history";
}

export default {
  name: "Home",
  components: {
    VCloseFormButton,
    VModal: () => import("@/components/UI/VModal.vue"),
    MultipleResults: () => import("@/components/MultipleResults.vue"),
    VScanner,
    PulseLoader,
  },
  mixins: [checkAdmin],
  data() {
    return {
      scanner: false,
      searchTerm: "",
      showNotFound: false,
      dialogOptions: {},
      loading: false,
      showArchived: false,
    };
  },
  methods: {
    ...mapMutations(["setSearchResults", "setCurrentInstrument", "setNewInstrumentNumber"]),
    detected(result) {
      if (result.codeResult.code !== this.searchTerm) {
        this.searchTerm = result.codeResult.code;
        this.scanner = false;
        this.onSubmit();
      }
    },
    async onSubmit() {
      if (this.loading) {
        return;
      }
      const path = getPath(this.searchTerm);
      try {
        this.loading = true;
        const response = await API.post("instrument-inventory", path, { body: { term: this.searchTerm } });
        this.loading = false;
        if (path === "search/number" || this.showArchived) {
          this.setSearchResults(response);
        } else {
          this.setSearchResults(response.filter(item => !item.archived));
        }
        if (this.searchResults.length === 1) {
          this.$toasted.info("Instrument Found", { duration: 2000 });
          this.setCurrentInstrument(this.searchResults[0]);
        } else if (this.searchResults.length < 1) {
          this.$toasted.info("Only archived instruments found", {
            duration: 2000,
          });
        } else {
          this.$toasted.info("Multiple Instruments Found", { duration: 2000 });
        }
      } catch (e) {
        this.loading = false;
        if (e.response.status === 404) {
          if (path === "search/number") {
            this.showNotFound = true;
          } else {
            // TODO: configure toasted so it doesn't suck
            this.$toasted.error(`Could not find instrument assigned to ${this.searchTerm}`, { duration: 2000 });
          }
        } else {
          this.$toasted.error(`Error: ${e.response.data}`, { duration: 2000 });
        }
      }
    },
    beginCreateInstrument() {
      this.setNewInstrumentNumber(this.searchTerm);
      this.$router.push("/new");
    },
    showArchivedResults(results) {
      this.setSearchResults(results);
    },
  },
  computed: mapState(["searchResults"]),
};
</script>

