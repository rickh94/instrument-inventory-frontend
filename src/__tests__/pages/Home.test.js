import React from 'react'
import Home from '../../pages/Home'
import { render, cleanup } from '../../test-utils'
import { API } from 'aws-amplify'

beforeEach(cleanup)
beforeAll(() => {
  console.error = jest.fn()
})

describe('<Home />', () => {
  test('matches snapshot', () => {
    API.get = () => Promise.resolve([])
    const { container } = render(<Home history={{ push: jest.fn() }} />)
    expect(container).toMatchSnapshot()
  })

  test('renders find instrument', () => {
    API.get = () => Promise.resolve([])
    const { queryByText } = render(<Home history={{ push: jest.fn() }} />)

    expect(queryByText('Find an Instrument')).toBeTruthy()
  })

  test('renders todo list', () => {
    API.get = () => Promise.resolve([])
    const { queryByText } = render(<Home history={{ push: jest.fn() }} />)

    expect(queryByText('Todo List')).toBeTruthy()
  })
})
