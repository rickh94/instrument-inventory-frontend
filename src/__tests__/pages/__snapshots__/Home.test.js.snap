// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 WithStyles(ForwardRef(Paper))-root-3 MuiPaper-rounded"
  >
    <div
      class="makeStyles-heading-33"
      data-testid="loading-header"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
      >
        Find an Instrument
      </h5>
    </div>
    <form>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          data-testid="label"
          for="scanner-field"
        >
          Name or Number
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          data-testid="input-field"
          required=""
        >
          <input
            aria-describedby="scanner-error"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            id="scanner-field"
            required=""
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              data-testid="scan-button"
              tabindex="0"
              title="Scan Barcode"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-barcode fa-w-16 "
                  data-icon="barcode"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 448V64h18v384H0zm26.857-.273V64H36v383.727h-9.143zm27.143 0V64h8.857v383.727H54zm44.857 0V64h8.857v383.727h-8.857zm36 0V64h17.714v383.727h-17.714zm44.857 0V64h8.857v383.727h-8.857zm18 0V64h8.857v383.727h-8.857zm18 0V64h8.857v383.727h-8.857zm35.715 0V64h18v383.727h-18zm44.857 0V64h18v383.727h-18zm35.999 0V64h18.001v383.727h-18.001zm36.001 0V64h18.001v383.727h-18.001zm26.857 0V64h18v383.727h-18zm45.143 0V64h26.857v383.727h-26.857zm35.714 0V64h9.143v383.727H476zm18 .273V64h18v384h-18z"
                    fill="currentColor"
                  />
                </svg>
                <span
                  class="MuiTypography-root MuiTypography-srOnly"
                  data-testid="sr-text"
                >
                  Scan Barcode
                </span>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        
      </div>
      <div
        class="MuiFormGroup-root MuiFormGroup-row"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-lastButton-31 MuiButton-text"
          data-testid="clear-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Clear
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-firstButton-32 MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
          data-testid="submit-button"
          disabled=""
          tabindex="-1"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Submit
          </span>
        </button>
      </div>
    </form>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 WithStyles(ForwardRef(Paper))-root-3 MuiPaper-rounded"
  >
    <div
      class="makeStyles-heading-33"
      data-testid="loading-header"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
      >
        Todo List
      </h5>
    </div>
    <br />
    <table
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
            scope="col"
          >
            Status
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
            scope="col"
          >
            Task
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
            scope="col"
          >
            Instrument
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
            scope="col"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-hover"
        >
          <td
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              style="margin-left: 3px;"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            Create Todo
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          />
          <td
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          />
        </tr>
        <tr
          class="MuiTableRow-root MuiTableRow-hover"
          data-testid="toggle-completed"
        >
          <td
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
              />
            </svg>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            Show 
            Completed
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          />
          <td
            class="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          />
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="makeStyles-footer-1"
  >
    <p
      class="MuiTypography-root makeStyles-footerItem-2 MuiTypography-body2 MuiTypography-colorInherit"
    >
      Â© Rick Henry 2019 |
       
    </p>
    <a
      class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
      href="http://chittagongit.com/icon/violin-icon-27.html"
      rel="noopener"
      target="_blank"
    >
      <p
        class="MuiTypography-root makeStyles-footerItem-2 MuiTypography-body2 MuiTypography-colorInherit"
      >
         
        Icon Credit
      </p>
    </a>
  </div>
</div>
`;
